### PR TITLE
Fix data race with responses using gRPC

### PR DESCRIPTION
**Pull request by peter-edge**
https://github.com/yarpc/yarpc-go/issues/1415

See the comment in the file:

```
We need to copy the bytes out so that we do not use the buffer-pool-backed byte slices.
grpc does a send back to the client on a loop in a separate goroutine, which may occur
after we return from this function. Because our codec is a passthrough, the "marshalled"
data from the interface{} given to SendMsg is the original byte buffer from the buffer pool,
which grpc can modify. Without this, the test TestDataRace in integration_test.go
will be set off.
```

I looked at other cases in `transport/grpc` and I think this is the only occurrence. There are potentially cleaner solutions but this fixes the problem for now

### CODE REVIEW

**Comment by prashantv** [comment]
> Because our codec is a passthrough, the "marshalled"
data from the interface{} given to SendMsg is the original byte buffer from the buffer pool,
which grpc can modify

I'm a little confused -- gRPC can modify the buffer we return from the codec's `Marshal`? I don't think that's the case here. I don't see anything modifying the data slice that's passed in to the underlying writer,
https://github.com/grpc/grpc-go/blob/7f2472bbc6ac269af51a9c47ae8ef8c625383e1f/transport/http2_client.go#L644

The `SendMsg` queues up the write on a channel:
https://github.com/grpc/grpc-go/blob/7f2472bbc6ac269af51a9c47ae8ef8c625383e1f/transport/http2_client.go#L718
`// controlBuffer is an unbounded channel of item.`

This is read from a background goroutine, 
https://github.com/grpc/grpc-go/blob/7f2472bbc6ac269af51a9c47ae8ef8c625383e1f/transport/http2_client.go#L305

The problem is that we were releasing the buffer after `Write`, but the buffer is accessed (for reads) in a background goroutine.

We were subject to the same data race that we ran into with thriftrw for gRPC, except it affected every read/write.

We should remove the pool completely, I don't think this half pooled then copy approach gives us any performance benefit.

**Comment by peter-edge** [comment]
You're looking at the client, not the server. If you trace the code that happens after `SendMsg`, you'll see the issue, that there are reads after `SendMsg`, correct.

I can remove the pooling completely, I mentioned this in the chat room, this might be the preferred approach.

**Comment by kriskowal** [comment]
Naïve question: would you say that copying bytes out of a buffer pool into unpooled byte allocations nullifies the advantage of the buffer pool?

I’m in favor of your suggestion to remove the pool completely. If we need it to claw back performance, it might hasten the time to cut out the grpc library that stands between us and http2.

**Comment by willhug** [comment]
(removing the pool also works)

File: transport/grpc/response_writer.go [hunk 1]
```diff
@@ -41,7 +40,7 @@ func newResponseWriter() *responseWriter {
 
 func (r *responseWriter) Write(p []byte) (int, error) {
 	if r.buffer == nil {
-		r.buffer = bufferpool.Get()
+		r.buffer = bytes.NewBuffer(make([]byte, 0, 1024))
```
Comment (prashantv) [inline transport/grpc/response_writer.go]: can you do `len(p)`

Comment (willhug) [reply transport/grpc/response_writer.go]: done

File: transport/grpc/handler.go [hunk 2]
```diff
@@ -180,8 +180,16 @@ func (h *handler) handleUnary(
 	err := h.handleUnaryBeforeErrorConversion(ctx, transportRequest, responseWriter, start, handler)
 	err = handlerErrorToGRPCError(err, responseWriter)
 
+	// We need to copy the bytes out so that we do not use the buffer-pool-backed byte slices.
```
Comment (prashantv) [inline transport/grpc/handler.go]: i don't think this comment makes sense anymore? we can revert this change right?

File: transport/grpc/integration_test.go [hunk 3]
```diff
@@ -170,6 +170,17 @@ func TestYARPCMaxMsgSize(t *testing.T) {
 	})
 }
 
+func TestDataRace(t *testing.T) {
```
Comment (prashantv) [inline transport/grpc/integration_test.go]: does this test fail without the fix?

Comment (willhug) [reply transport/grpc/integration_test.go]: It does not...  I don't think we'll be able to get significant pooling tests without #1408, I'll leave the test, but it's mostly just "large" test.
